vowpalwabbit/arguments.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <typeinfo>
+#include <memory>
+
+namespace VW {
+
+  struct base_argument {
+    base_argument(std::string name, size_t type_hash)
+      : m_name(name), m_type_hash(type_hash)
+    {}
+
+    std::string m_name;
+    size_t m_type_hash;
+    std::string m_help = "";
+    std::string m_short_name = "";
+    bool m_keep = false;
+
+    virtual ~base_argument() {}
+  };
+
+  template<typename T>
+  struct typed_argument : base_argument {
+    typed_argument(std::string name, T* location)
+      : base_argument(name, typeid(T).hash_code()) {
+      m_locations.push_back(location);
+    }
+
+    static size_t type_hash() {
+      return typeid(T).hash_code();
+    }
+
+    typed_argument& default_value(T value) {
+      m_default_value = std::make_shared<T>(value);
+      return *this;
+    }
+
+    bool default_value_supplied() {
+      return m_default_value.get() != nullptr;
+    }
+
+    T default_value() {
+      return m_default_value ? *m_default_value : T();
+    }
+
+    typed_argument& short_name(std::string short_name) {
+      m_short_name = short_name;
+      return *this;
+    }
+
+    typed_argument& help(std::string help) {
+      m_help = help; return *this;
+    }
+
+    typed_argument& keep(bool keep = true) {
+      m_keep = keep; return *this;
+    }
+
+    bool value_supplied() {
+      return m_value.get() != nullptr;
+    }
+
+    typed_argument& value(T& value) {
```

Comment:
Should setting this also set the value for all of m_locations. This sets a dangerous precedent about event based parameter/variable synchronization in the library though.

Reply:
I'm not quite following this.

Reply:
Because of the fact we pass pointers to the location to write the value upon parsing, providing a function to specify the stored value begs the question of whether or not when giving this value if it is also written to the location pointers

Reply:
No?  It seems like all writes should happen at parsing time.

Reply:
For consistency between defaults and non-defaults.

Reply:
I agree with you, all writes should happen at parsing time